Flaky `go.etcd.io/etcd/server/v3/etcdserver/api/v3compactor: TestPeriodicHourly `

### Which Github Action / Prow Jobs are flaking?

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/etcd-io_etcd/19426/pull-etcd-coverage-report/1895067242123497472

```

go.etcd.io/etcd/server/v3/etcdserver/api/v3compactor: TestPeriodicHourly expand_less | 1s
-- | --
{Failed  === RUN   TestPeriodicHourly     logger.go:146: 2025-02-27T11:08:18.530Z	INFO	starting auto periodic compaction	{"revision": 1, "compact-period": "2h0m0s"}     logger.go:146: 2025-02-27T11:08:18.530Z	INFO	completed auto periodic compaction	{"revision": 1, "compact-period": "2h0m0s", "took": "0s"}     logger.go:146: 2025-02-27T11:08:18.850Z	INFO	starting auto periodic compaction	{"revision": 10, "compact-period": "2h0m0s"}     logger.go:146: 2025-02-27T11:08:18.851Z	INFO	completed auto periodic compaction	{"revision": 10, "compact-period": "2h0m0s", "took": "0s"}     logger.go:146: 2025-02-27T11:08:19.032Z	INFO	starting auto periodic compaction	{"revision": 20, "compact-period": "2h0m0s"}     logger.go:146: 2025-02-27T11:08:19.032Z	INFO	completed auto periodic compaction	{"revision": 20, "compact-period": "2h0m0s", "took": "0s"}     logger.go:146: 2025-02-27T11:08:19.281Z	INFO	starting auto periodic compaction	{"revision": 30, "compact-period": "2h0m0s"}     periodic_test.go:72: len(actions) = 0, expected >= 1 --- FAIL: TestPeriodicHourly (1.49s) }


```

FYI. Previously I raised a similar issue https://github.com/etcd-io/etcd/issues/17054, which was resolved in https://github.com/etcd-io/etcd/pull/18608. It seems the fix still has flaw.

### Which tests are flaking?

.

### Github Action / Prow Job link

_No response_

### Reason for failure (if possible)

_No response_

### Anything else we need to know?

_No response_

--- a/server/etcdserver/api/v3compactor/periodic_test.go
+++ b/server/etcdserver/api/v3compactor/periodic_test.go
@@ -16,6 +16,7 @@ package v3compactor
 
 import (
 	"errors"
+	"fmt"
 	"reflect"
 	"testing"
 	"time"
@@ -49,7 +50,7 @@ func TestPeriodicHourly(t *testing.T) {
 	}
 
 	// very first compaction
-	a, err := compactable.Wait(1)
+	a, err := waitWithRetry(t, compactable)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -67,7 +68,7 @@ func TestPeriodicHourly(t *testing.T) {
 			fc.Advance(tb.getRetryInterval())
 		}
 
-		a, err = compactable.Wait(1)
+		a, err = waitWithRetry(t, compactable)
 		if err != nil {
 			t.Fatal(err)
 		}
@@ -100,7 +101,7 @@ func TestPeriodicMinutes(t *testing.T) {
 	}
 
 	// very first compaction
-	a, err := compactable.Wait(1)
+	a, err := waitWithRetry(t, compactable)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -117,7 +118,7 @@ func TestPeriodicMinutes(t *testing.T) {
 			fc.Advance(tb.getRetryInterval())
 		}
 
-		a, err := compactable.Wait(1)
+		a, err := waitWithRetry(t, compactable)
 		if err != nil {
 			t.Fatal(err)
 		}
@@ -162,7 +163,7 @@ func TestPeriodicPause(t *testing.T) {
 	fc.Advance(tb.getRetryInterval())
 
 	// T=3h6m
-	a, err := compactable.Wait(1)
+	a, err := waitWithRetry(t, compactable)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -197,7 +198,7 @@ func TestPeriodicSkipRevNotChange(t *testing.T) {
 	}
 
 	// very first compaction
-	a, err := compactable.Wait(1)
+	a, err := waitWithRetry(t, compactable)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -228,7 +229,7 @@ func TestPeriodicSkipRevNotChange(t *testing.T) {
 		fc.Advance(tb.getRetryInterval())
 	}
 
-	a, err = compactable.Wait(1)
+	a, err = waitWithRetry(t, compactable)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -243,4 +244,24 @@ func waitOneAction(t *testing.T, r testutil.Recorder) {
 	if actions, _ := r.Wait(1); len(actions) != 1 {
 		t.Errorf("expect 1 action, got %v instead", len(actions))
 	}
+}
+
+func waitWithRetry(t *testing.T, compactable *fakeCompactable) ([]testutil.Action, error) {
+	t.Helper()
+
+	var lastErr error
+	var actions []testutil.Action
+
+	expectedActions, maxRetries := 1, 5
+	for retry := 0; retry < maxRetries; retry++ {
+		actions, lastErr = compactable.Wait(expectedActions)
+		if lastErr == nil || len(actions) >= expectedActions {
+			return actions, nil
+		}
+		// Exponential backoff
+		backoffTime := time.Duration(10*(1<<retry)) * time.Millisecond
+		t.Logf("Retry %d: waiting %v before next attempt (last error: %v)", retry+1, backoffTime, lastErr)
+		time.Sleep(backoffTime)
+	}
+	return nil, fmt.Errorf("after %d retries, last error: %w", maxRetries, lastErr)
 }

